### PR TITLE
Added plot library

### PR DESCRIPTION
**Pull request by alexandersundin**

### CODE REVIEW

File: source/math/FloatPoint2D.ooc [hunk 1]
```diff
@@ -40,8 +40,8 @@ FloatPoint2D: cover {
 	round: func -> This { This new(this x round(), this y round()) }
 	ceiling: func -> This { This new(this x ceil(), this y ceil()) }
 	floor: func -> This { This new(this x floor(), this y floor()) }
-	minimum: func (ceiling: This) -> This { This new(this x minimum(ceiling x), this y minimum(ceiling y)) }
-	maximum: func (floor: This) -> This { This new(this x maximum(floor x), this y maximum(floor y)) }
+	minimum: func (ceiling: This) -> This { This new(Float minimum(this x, ceiling x), Float minimum(this y, ceiling y)) }
+	maximum: func (floor: This) -> This { This new(Float maximum(this x, floor x), Float maximum(this y, floor y)) }
```
Comment (davidhesselbom) [inline source/math/FloatPoint2D.ooc]: What was the behavior of this function prior to this change, and how does the behavior change with this modification?

File: source/plot/Axis.ooc [hunk 2]
```diff
@@ -0,0 +1,170 @@
+use ooc-math
+import svg/Shapes
+import math
+
+Orientation: enum {
+	HORIZONTAL
+	VERTICAL
```
Comment (davidhesselbom) [inline source/plot/Axis.ooc]: Enums are written like class names, so `Horizontal` and `Vertical`, respectively.

File: source/plot/PlotData2D.ooc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+use ooc-math
+
+PlotData2D: abstract class {
+	label: String { get set}
+	fontSize: Int { get set }
+	lineWidth: Float { get set }
+	colorBgra: ColorBgra { get set }
+	color: String { get set }
+	opacity: Float { get set }
+	dataSerie: VectorList<FloatPoint2D> { get set }
```
Comment (davidhesselbom) [inline source/plot/PlotData2D.ooc]: "Series" is "series" both in singular and plural, so `dataSeries`.

File: source/plot/PlotData2D.ooc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+	init: func ~default() {
+		this init(VectorList<FloatPoint2D> new(), "")
+	}
+	init: func ~dataserie(dataSerie: VectorList<FloatPoint2D>) {
+		this dataSerie = dataSerie
+		this lineWidth = 1
+		this fontSize = 14
+		if (this label == null)
+			this label = ""
```
Comment (davidhesselbom) [inline source/plot/PlotData2D.ooc]: If this is a constructor, how is label ever going to be anything but `null` before being assigned?

**Comment by davidhesselbom** [comment]
Go to `ooc-kean/test` and run `rock -v -r -O3 --gc=off -lpthread plot/SVGPlotTest.ooc` followed by `valgrind --leak-check=full --log-file=memcheck.valgrindlog ./SVGPlotTest`.

I get 
```
==23820== LEAK SUMMARY:
==23820==    definitely lost: 835,264 bytes in 35,105 blocks
==23820==    indirectly lost: 32,210,608 bytes in 70,047 blocks
==23820==      possibly lost: 162,696 bytes in 2 blocks
==23820==    still reachable: 84,664 bytes in 1,107 blocks
==23820==         suppressed: 0 bytes in 0 blocks
```

File: ooc-plot.use [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Name: OOC Plot
```
Comment (davidhesselbom) [inline ooc-plot.use]: It's ooc, not OOC.

File: source/draw/Color.ooc [hunk 6]
```diff
@@ -173,6 +173,10 @@ ColorBgr: cover {
 	operator != (other: ColorYuv) -> Bool { !this equals(other) }
 	operator != (other: This) -> Bool { !this equals(other) }
 	operator != (other: ColorBgra) -> Bool { !this equals(other) }
+	svgRGBToString: func() -> String {
+		result:= this red toString() + "," + this green toString() + "," + this blue toString()
```
Comment (davidhesselbom) [inline source/draw/Color.ooc]: `:=` should always be surrounded by space on both sides.

File: source/plot/Axis.ooc [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+			}
+		}
+	}
+	orientation: Orientation { get set }
+	fontSize: Int { get set }
+	gridOn: Bool { get set }
+	precision: Int { get set }
+	roundAxisEndpoints: Bool { get set }
+
+	init: func(=orientation) {
```
Comment (davidhesselbom) [inline source/plot/Axis.ooc]: `func` is also surrounded by space on both sides.

File: source/plot/SVGPlot.ooc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+	init: func ~datasets (datasets: VectorList<PlotData2D>) {
+		this init(datasets, "")
+	}
+
+	init: func ~datasetTitle (dataset: PlotData2D, title: String) {
+		this datasets = VectorList<PlotData2D> new()
+		this datasets add(dataset)
+	this init()
+	this title = title
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: Indentation

File: source/plot/SVGPlot.ooc [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	init: func ~datasetsTitle (datasets: VectorList<PlotData2D>, title: String) {
+		this datasets = datasets
+	this init()
+	this title = title
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: Indentation

File: source/plot/SVGPlot.ooc [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+		result += this xAxis getSVG(plotAreaSize, xAxisSize, bottomLeftCornerOfPlot, translationToRealOrigo, scaling)
+		result += this yAxis getSVG(plotAreaSize, yAxisSize, bottomLeftCornerOfPlot, translationToRealOrigo, scaling)
+		result += "<rect desc='Plot-border' x='" + bottomLeftCornerOfPlot x toString() + "' y='" + (bottomLeftCornerOfPlot y - plotAreaSize y) toString() + "' width='" + plotAreaSize x toString() + "' height='" + plotAreaSize y toString() + "' stroke='black' fill='none'/>"
+		result += "<svg desc='Data' x='" + bottomLeftCornerOfPlot x toString() + "' y='" + (bottomLeftCornerOfPlot y - plotAreaSize y) toString() + "' width='" + plotAreaSize x toString() + "' height='" + plotAreaSize y toString() + "'>"
+		result += "<g transform='translate(" + translationToRealOrigo toString() + ")'>"
+		if (!this datasets empty()) {
+			for (i in 0..this datasets count) {
+				result += this datasets[i] getSVG(scaling)
+			}
+		}
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: These brackets are both unnecessary:
```ooc
if (!this datasets empty())
    for (i in 0..this datasets count)
        result += this datasets[i] getSVG(scaling)
```
works just as well.

File: source/plot/SVGPlot.ooc [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	setAxesMinMax: func() {
+		if (!datasets empty()) {
+			min:= datasets[0] minValues()
+			max:= datasets[0] maxValues()
+
+			for (i in 0..datasets count) {
+					min = min minimum(datasets[i] minValues())
+					max = max maximum(datasets[i] maxValues())
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: Indentation

File: source/plot/SVGPlot.ooc [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+			this yAxis min = min y
+			this yAxis max = max y
+		}
+	}
+
+	setColor: func () {
+		if (!datasets empty()) {
+			noneColor:= ColorBgra new(0,0,0,0)
+			for (j in 0..datasets count) {
+				if(!(noneColor == datasets[j] colorBgra)){
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: `if (`, `)) {`

File: source/plot/SVGPlot.ooc [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+					datasets[j] color = datasets[j] colorBgra svgRGBToString()
+					datasets[j] opacity = ((datasets[j] colorBgra svgRGBAlpha()) as Float) / 255.0f
+				}else{
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: `} else {`

File: source/plot/SVGPlot.ooc [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+					datasets[j] color = colorList[this colorCount % this colorList count]
+					datasets[j] opacity = 1
+					this colorCount+=1
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: ` += 1`

**Comment by davidhesselbom** [comment]
The svg output code is confined to a single line. Add newlines where appropriate.

File: source/plot/LinePlotData2D.ooc [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+	init: func ~label(dataSerie: VectorList<FloatPoint2D>, label: String) {
+		super(dataSerie, label)
+	}
+
+	init: func ~color(dataSerie: VectorList<FloatPoint2D>, colorBgra: ColorBgra) {
+		super(dataSerie, colorBgra)
+	}
+
+	init: func ~labelColor(dataSerie: VectorList<FloatPoint2D>, label: String, colorBgra: ColorBgra) {
+		super(dataSerie,label,colorBgra)
```
Comment (davidhesselbom) [inline source/plot/LinePlotData2D.ooc]: Leave spaces between parameters: `super(dataSerie, label, colorBgra)`

File: source/plot/LinePlotData2D.ooc [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	getSVG: func(scaling: FloatPoint2D) -> String {
+		result:= ""
+		if (!this dataSerie empty()) {
+			result += "<path stroke='" + this color + "' stroke-opacity='" + this opacity toString() + "' fill='none' stroke-width='" + this lineWidth toString() + "' d='M " + (scaling x * this dataSerie[0] x) toString() + " " + (- scaling y * this dataSerie[0] y) toString() + " L "
+			for (j in 1..this dataSerie count) {
+				result += (scaling x * this dataSerie[j] x) toString() + " " + (- scaling y * this dataSerie[j] y) toString() + " "
+			}
```
Comment (davidhesselbom) [inline source/plot/LinePlotData2D.ooc]: Needless `{ }`

File: source/math/FloatPoint2D.ooc (hunk 1 above)
Comment (alexandersundin) [inline source/math/FloatPoint2D.ooc]: Before the changes the methods returned the second parameter, calling minimum(a,b) would return b, regardless of the x- and y-magnitudes of the parameters. With the changes the methods return a FloatPoint2D consisting of the largest/smallest magnitude of x and y .

Example:
Given the points a=(-1,1) and b=(1,-1), calling minimum(a,b) would give the following result:
Before changes: (1,-1)
After changes: (-1,-1)

Comment (davidhesselbom) [inline source/math/FloatPoint2D.ooc]: Interesting...

**Comment by alexandersundin** [comment]
Updated

File: source/draw/Color.ooc [hunk 17]
```diff
@@ -214,6 +218,14 @@ ColorBgra: cover {
 	operator != (other: ColorYuv) -> Bool { !this equals(other) }
 	operator != (other: ColorBgr) -> Bool { !this equals(other) }
 	operator != (other: This) -> Bool { !this equals(other) }
+	svgRGBToString: func() -> String {
+		result:= "rgb(" clone() & this bgr svgRGBToString() & ")" clone()
```
Comment (davidhesselbom) [inline source/draw/Color.ooc]: Always leave spaces around `:=`

File: source/plot/Axis.ooc [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+		result := valueAndPower[0]
+		if (valueAndPower[0] != 0) {
+			result = up ? ceil(valueAndPower[0]) : floor(valueAndPower[0])
+			result *= valueAndPower[1]
+		}
+		gc_free(valueAndPower data)
+		result
+	}
+
+	decomposeToValueAndPower: func (value: Float, valueDigits: Int) -> Float[] {
```
Comment (davidhesselbom) [inline source/plot/Axis.ooc]: Wouldn't it be better to use a [tuple](http://ooc-lang.org/docs/lang/tuples/)?

File: source/plot/Axis.ooc [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+	roundToValueDigits: func (value: Float, valueDigits: Int, up: Bool) -> Float {
+		valueAndPower := this decomposeToValueAndPower(value, valueDigits)
+		result := valueAndPower[0]
+		if (valueAndPower[0] != 0) {
+			result = up ? ceil(valueAndPower[0]) : floor(valueAndPower[0])
+			result *= valueAndPower[1]
+		}
+		gc_free(valueAndPower data)
+		result
+	}
```
Comment (davidhesselbom) [inline source/plot/Axis.ooc]: Pretty sure this function should be in `FloatExtension.ooc`, not particular to the Plot library.

File: source/plot/Axis.ooc [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+			power += 1
+			value /= 10
+		}
+		while (value <= 1) {
+			power -= 1
+			value *= 10
+		}
+		result := "10^" clone() & power toString()
+		result
+	}
```
Comment (davidhesselbom) [inline source/plot/Axis.ooc]: Same here, these last 4 functions seem generic to Floats and shouldn't be specific to the Plot library.

File: source/plot/PlotData2D.ooc [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+	init: func ~color(dataSeries: VectorList<FloatPoint2D>, colorBgra: ColorBgra) {
+		this colorBgra = colorBgra
+		this init(dataSeries)
+	}
+
+	init: func ~labelColor(dataSeries: VectorList<FloatPoint2D>, label: String, colorBgra: ColorBgra) {
+		this label = label
+		this colorBgra = colorBgra
+		this init(dataSeries)
+	}
```
Comment (davidhesselbom) [inline source/plot/PlotData2D.ooc]: ooc supports parameter defaults, so you could reduce this to
```ooc
init: func ~default() {
    this init(VectorList<FloatPoint2D> new())
}
init: func ~dataserie(=dataSeries, label := "", colorBgra := ColorBgra new()) {
    this lineWidth = 1
    this fontSize = 14
    this label = label
    this colorBgra = colorBgra
}
init: func ~color(dataSeries: VectorList<FloatPoint2D>, colorBgra: ColorBgra) {
    this init(dataSeries, "", colorBgra)
}
```
... I think.

File: source/plot/PlotData2D.ooc [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+	lineWidth: Float { get set }
+	colorBgra: ColorBgra { get set }
+	color: String { get set }
+	opacity: Float { get set }
+	dataSeries: VectorList<FloatPoint2D> { get set }
+
+	init: func ~default() {
+		this init(VectorList<FloatPoint2D> new(), "")
+	}
+	init: func ~dataserie(dataSeries: VectorList<FloatPoint2D>) {
```
Comment (davidhesselbom) [inline source/plot/PlotData2D.ooc]: dataSeries

File: source/plot/SVGPlot.ooc [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+	yAxis: Axis { get set }
+	colorList: VectorList<String> { get set }
+	colorCount := 0
+	plotAreaPercentage: Float
+	symmetric: Bool
+
+
+	init: func {
+		this plotAreaPercentage = 0.8f
+		this title=""
```
Comment (davidhesselbom) [inline source/plot/SVGPlot.ooc]: leave spaces around `=`

File: source/plot/SVGWriter2D.ooc [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+		fileWriter close()
+		output free()
+		fileWriter free()
+	}
+
+	prepareOutput: func -> String {
+		result := "<?xml version='1.0' standalone='no'?>\n
+<!DOCTYPE svg PUBLIC '-//W3C//DTD SVG 1.1//EN'
+	'http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd'>
+	<svg xmlns:svg='http://www.w3.org/2000/svg' xmlns='http://www.w3.org/2000/svg' version='1.1' width='" clone() & this width toString() & "' height='" clone() & this height toString() & "'>\n" clone()
```
Comment (davidhesselbom) [inline source/plot/SVGWriter2D.ooc]: This looks rather messy...

File: source/plot/SVGWriter2D.ooc [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+		if (!this svgPlots empty()) {
+			numPlotsX: Int
+			numPlotsY: Int
+			if (this svgPlots count > 1 && this svgPlots count < 5) {
+				numPlotsX = 2
+			} else if (this svgPlots count >= 5) {
+				numPlotsX = 3
+			} else {
+				numPlotsX = 1
+			}
```
Comment (davidhesselbom) [inline source/plot/SVGWriter2D.ooc]: Needless `{ }`

**Comment by alexandersundin** [comment]
Updated

File: source/plot/Axis.ooc [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+				position y += - scaling y * tick
+			}
+		}
+		result = result & "</g>\n" clone()
+		result
+	}
+
+	getFirstTickValue: func -> Float {
+		result: Float
+		if ((this min < this tick && this min > 0.0f) || (this min > this tick && this min > 0.0f && this min < 1.0f) )
```
Comment (davidhesselbom) [inline source/plot/Axis.ooc]: `) )`?

**Comment by alexandersundin** [comment]
Updated

File: source/plot/ScatterPlotData2D.ooc [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+import svg/Shapes
+
+Shape: enum {
+	CIRCLE
+	SQUARE
+}
+
+ScatterPlotData2D: class extends PlotData2D {
+	shape := Shape CIRCLE
+	scalingRelativeLineWidth:= 5.0f
```
Comment (davidhesselbom) [inline source/plot/ScatterPlotData2D.ooc]: space around `:=`

File: source/plot/ScatterPlotData2D.ooc [hunk 28]
```diff
@@ -0,0 +1,63 @@
+use ooc-collections
+use ooc-math
+use ooc-draw
+import PlotData2D
+import svg/Shapes
+
+Shape: enum {
+	CIRCLE
+	SQUARE
```
Comment (davidhesselbom) [inline source/plot/ScatterPlotData2D.ooc]: Enums are written like class names: `Circle Square`

File: source/plot/ScatterPlotData2D.ooc [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+					case Shape CIRCLE =>
+						r := this scalingRelativeLineWidth / 2.0f * this lineWidth
+						result = result & Shapes circle(scaling x * this dataSeries[i] x, - scaling y * this dataSeries[i] y, r, this opacity, this color)
+
+					case Shape SQUARE =>
+						x := scaling x * this dataSeries[i] x - this scalingRelativeLineWidth / 2.0f * this lineWidth
+						y := -scaling y * this dataSeries[i] y - this scalingRelativeLineWidth / 2.0f * this lineWidth
+						width := this scalingRelativeLineWidth * this lineWidth
+						result = result & Shapes rect(x, y, width, width, this opacity, this color)
+				}
```
Comment (davidhesselbom) [inline source/plot/ScatterPlotData2D.ooc]: And if it's neither Circle nor Square, then what?

File: source/plot/ScatterPlotData2D.ooc [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+	init: func ~color(dataSeries: VectorList<FloatPoint2D>, colorBgra: ColorBgra) {
+			super(dataSeries, "", colorBgra)
+	}
+
+	getSVG: func (scaling: FloatPoint2D) -> String {
+		result := ""
+		if (!this dataSeries empty()) {
+			for (i in 0..this dataSeries count) {
+				match (this shape) {
+
```
Comment (davidhesselbom) [inline source/plot/ScatterPlotData2D.ooc]: Not big on this newline here

File: source/plot/SVGWriter2D.ooc [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+		output := prepareOutput()
+		fileWriter := FileWriter new(this filename, false)
+		fileWriter write(output)
+		fileWriter close()
+		output free()
+		fileWriter free()
+	}
+
+	prepareOutput: func -> String {
+		result := "<?xml version='1.0' standalone='no'?>\n"
```
Comment (davidhesselbom) [inline source/plot/SVGWriter2D.ooc]: So this line needs not be cloned, but the rest do? Fascinating...

File: source/plot/ScatterPlotData2D.ooc (hunk 29 above)
Comment (alexandersundin) [inline source/plot/ScatterPlotData2D.ooc]: Then it is an undefined shape and the data will not be printed. Does that seem ok?

**Comment by alexandersundin** [comment]
Updated

File: source/plot/ScatterPlotData2D.ooc (hunk 29 above)
Comment (davidhesselbom) [inline source/plot/ScatterPlotData2D.ooc]: In the very next function, you do
```ooc
symbol: String
...
case => symbol = ""
```
which seems like a better idea.

File: source/plot/svg/Shapes.ooc [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+		"<rect x='" clone() & x toString() & "' y='" clone() & y toString() & "' width='" clone() & width toString() & "' height='" clone() & height toString() & "' fill='" clone() & color clone() & "' fill-opacity='" clone() & opacity toString() & "'/>\n" clone()
+	}
+
+	text: static func ~simple(position: FloatPoint2D, text: String, fontSize: Int) -> String {
+		This text(position, text, fontSize, "start")
+	}
+
+	text: static func ~textAnchor(position: FloatPoint2D, text: String, fontSize: Int, textAnchor: String) -> String {
+		"<text id='" << (text + "' x='") & position x toString() & "' y='" clone() & position y toString() & "' font-size='" clone() & fontSize toString() & ("' fill='black'	text-anchor='" << (textAnchor << ("' >" << (text + "</text>\n"))))
+	}
```
Comment (davidhesselbom) [inline source/plot/svg/Shapes.ooc]: Lots of empty lines here...

**Comment by alexandersundin** [comment]
Updated

File: source/base/DateTime.ooc [hunk 33]
```diff
@@ -178,6 +185,15 @@ DateTime: cover {
...
   }
+  toFileSafeString: func -> String {
+    year := this _year toString()
+    month := this _month < 10 ? "0" << this _month toString() : this _month toString()
+    day := this _day < 10 ? "0" << this _day toString() : this _day toString()
+    hour := this _hour < 10 ? "0" << this _hour toString() : this _hour toString()
+    minute := this _minute < 10 ? "0" << this _minute toString() : this _minute toString()
+    second := this _second < 10 ? "0" << this _second toString() : this _second toString()
+    year >> "-" & month >> "-" & day >> "_" & hour >> "-" & minute >> "-" & second
+  }
```
Comment (davidhesselbom) [inline source/base/DateTime.ooc]: Fix tabs

File: source/plot/svg/Shapes.ooc [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+		"<rect x='" clone() & upperRight x toString() & "' y='" clone() & upperRight y toString() & "' width='" clone() & width toString() & "' height='" clone() & height toString() & "' fill='" clone() & color clone() & "' fill-opacity='" clone() & opacity toString() & "'/>\n" clone()
+	}
+	rect: static func ~Float(x, y, width, height, opacity: Float, color: String) -> String {
+		"<rect x='" clone() & x toString() & "' y='" clone() & y toString() & "' width='" clone() & width toString() & "' height='" clone() & height toString() & "' fill='" clone() & color clone() & "' fill-opacity='" clone() & opacity toString() & "'/>\n" clone()
+	}
+	text: static func ~simple(position: FloatPoint2D, text: String, fontSize: Int) -> String {
+		This text(position, text, fontSize, "start")
+	}
+	text: static func ~textAnchor(position: FloatPoint2D, text: String, fontSize: Int, textAnchor: String) -> String {
+		"<text id='" << (text + "' x='") & position x toString() & "' y='" clone() & position y toString() & "' font-size='" clone() & fontSize toString() & ("' fill='black'	text-anchor='" << (textAnchor << ("' >" << (text + "</text>\n"))))
```
Comment (davidhesselbom) [inline source/plot/svg/Shapes.ooc]: Fix the tab, and `"' >"` should be `"'>"`

**Comment by alexandersundin** [comment]
Updated

**Pull request merged**